crates/fuel-core/src/graphql_api/da_compression.rs

```diff
@@ -0,0 +1,193 @@
+                        .into_owned();
+                    match v {
+                        RegistryKeyspaceValue::$name(v) => Ok(v),
+                        _ => anyhow::bail!("Unexpected value in the registry"),
+                    }
+                }
+
+                fn write_registry(
+                    &mut self,
+                    key: fuel_core_types::fuel_compression::RegistryKey,
+                    value: $type,
+                ) -> anyhow::Result<()> {
+                    // Write the actual value
+                    self.db_tx
+                        .storage_as_mut::<DaCompressionTemporalRegistry>()
+                        .insert(&(RegistryKeyspace::$name, key), &RegistryKeyspaceValue::$name(value.clone()))?;
+
+                    // Remove the overwritten value from index, if any
+                    self.db_tx
+                        .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+                        .remove(&RegistryKeyspaceValue::$name(value.clone()))?;
+
+                    // Add the new value to the index
+                    self.db_tx
+                        .storage_as_mut::<DaCompressionTemporalRegistryIndex>()
+                        .insert(&RegistryKeyspaceValue::$name(value), &key)?;
+
+                    Ok(())
+                }
+
+                fn registry_index_lookup(
+                    &self,
+                    value: &$type,
+                ) -> anyhow::Result<Option<fuel_core_types::fuel_compression::RegistryKey>>
+                {
+                    Ok(self
+                        .db_tx
+                        .storage_as_ref::<DaCompressionTemporalRegistryIndex>()
+                        .get(&RegistryKeyspaceValue::$name(value.clone()))?
+                        .map(|v| v.into_owned()))
+                }
+            }
+
+        )*
+    };
+}
+
+// Arguments here should match the tables! macro from crates/compression/src/tables.rs
+impl_temporal_registry! {
+    address: Address,
+    asset_id: AssetId,
+    contract_id: ContractId,
+    script_code: ScriptCode,
+    predicate_code: PredicateCode,
+}
+
+impl<'a, Tx> UtxoIdToPointer for CompressTx<'a, Tx>
+where
+    Tx: OffChainDatabaseTransaction,
+{
+    fn lookup(
+        &self,
+        utxo_id: fuel_core_types::fuel_tx::UtxoId,
+    ) -> anyhow::Result<fuel_core_types::fuel_tx::CompressedUtxoId> {
+        for event in self.block_events {
+            match event {
+                Event::CoinCreated(coin) | Event::CoinConsumed(coin)
+                    if coin.utxo_id == utxo_id =>
+                {
+                    let output_index = coin.utxo_id.output_index();
+                    return Ok(fuel_core_types::fuel_tx::CompressedUtxoId {
+                        tx_pointer: coin.tx_pointer,
+                        output_index,
+                    });
+                }
+                _ => {}
+            }
+        }
+        panic!("UtxoId not found in the block events");
```

Comment:
We should return an error instead of panic, and maybe `debug_assert!(false)` if you want to catch it during local developing

Reply:
Maybe `unreachable!` instead of `panic!`? But I don't see why we should ever get here. What good does it do to return an error, this in unrecoverable anyway? If we cannot produce a compressed block, halting the chain is a reasonable action.